### PR TITLE
Split out TypeList into its own header and moved macros to Platform

### PR DESCRIPTION
**Pull request by Idclip**
Signed-off-by: Nick Avramoussis <4256455+Idclip@users.noreply.github.com>

### CODE REVIEW

**Review by jmlait — APPROVED** [review]

File: openvdb/openvdb/TypeList.h [hunk 1]
```diff
@@ -0,0 +1,531 @@
...
+
+template<typename... Ts> struct TypeList; // forward declaration
+
+namespace internal {
+
+// Implementation details of @c TypeList
+
+/// @brief Dummy struct, used as the return type from invalid or out-of-range
+///        @c TypeList queries.
+struct NullType {};
```
Comment (danrbailey) [inline openvdb/openvdb/TypeList.h:34]: While we're moving this about - could we rename the namespace from internal to something that's less likely to clash (typelist_internal?). 

Using `internal::NullType` is rather generic.

File: openvdb/openvdb/Platform.h [hunk 2]
```diff
@@ -229,4 +229,54 @@
     #define OPENVDB_HOUDINI_API OPENVDB_IMPORT
 #endif
 
+#if defined(__ICC)
```
Comment (danrbailey) [inline openvdb/openvdb/Platform.h:232]: Tempted to say we strip all of this out - we don't build or test for ICC so it's hard to know if any of this is still relevant. They're just suppressing warnings after all.

**Review by danrbailey — APPROVED** [review]

File: openvdb/openvdb/Platform.h (hunk 2 above)
Comment (Idclip) [reply openvdb/openvdb/Platform.h:232]: Some of these defines are still in use so no harm leaving them in for now

File: openvdb/openvdb/TypeList.h (hunk 1 above)
Comment (Idclip) [reply openvdb/openvdb/TypeList.h:34]: I've included `TypeList.h` back into `Types.h` so as not to to break anyone's code so doesn't really matter that it's being moved - but I'm happy to make this change

**Pull request merged**